[1.0.0-alpha2] application scopes and traits (#319)
oam-dev/spec: 5.application_scopes.md

Component A is isolated in its own network scope from components B, C, and D. This allows the infrastructure operator to supply different SDN settings for different groups of components, such as more restricted inbound/outbound rules on back-end components.

## Categories of application scope

There are two kinds of application scope types:

 - **Core application scope types**: Core application scope types define grouping constructs for basic runtime behavior. They have the following characteristics:
    - Core application scope types MUST be in the `core.oam.dev` namespace.
    - Core application scope types MUST all be supported by any implementation of the specification. They may be implemented by any available feature in the platform, but they MUST be implemented according to the specification of each core application scope type.

>     - Instances of core *workload* types MUST be deployed into an instance of each core application scope type. 

Reviewer: By "Instances of core *workload* types" are we referring to the workload CRs? Or are we referring to components that are configured to a certain kind of workload?

Author: It means workload CRs, I should fix it

Reviewer: lgtm

    - Runtimes MUST deploy each component instances into default "root" application scope instances when an application configuration does not otherwise specify an application scope for a component instance.

 - **Standard application scope types**: Standard application scope types MUST be defined in `standard.oam.dev` group. These application scopes provide operational functionality that is commonly used by application operators and implemented by most runtimes. Runtimes are RECOMMENDED to use the standard application scope definitions as defined in this specification when providing equivalent operational functionality to those listed in the standard application scope definitions. In other words, if a runtime is implementing application scope with behavior `foo` and a standard application scope definition for behavior `foo` exists, the runtime SHOULD use the standard `foo` application scope definition. Application operators that intend to maximize portability of their applications should use these application scope definitions when available. Although this does not _guarantee_ portability of an application, it is designed to increase portability across runtimes.

>  - **Extended application scope types**: Extended application scopes are optional per runtime, meaning that each runtime may choose which extended scopes are supported. It is expected that runtimes will implement extended scopes as they need. In this version of the spec, allowing user-defined extended application scope types is not supported.

Reviewer 2: ```suggestion
 - **Extended application scope types**:  Each runtime MAY choose to implement any number of extended application scopes. 
```

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `allowComponentOverlap` | `bool` | Y | | Determines whether a component is allowed to be in multiple instances of this scope type simultaneously. When false, the runtime implementation MUST produce an error and stop deployment if an attempt is made to place a component into more than one instance of this scope type simultaneously. |
| `definitionRef` | [`DefinitionRef`](#definitionRef) | Y | | Index of the scope definition. |

### DefinitionRef

We use the same [DefinitionRef](3.workload.md#DefinitionRef) in workload.

## Deployment

> Application scope instances are defined and deployed by their standalone way. Go to scopes folder to see [core scopes](scopes/core.md) and [extended scopes](scopes/extended.md) details.

Reviewer 2: Do we still have extended scopes definition?

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `allowComponentOverlap` | `bool` | Y | | Determines whether a component is allowed to be in multiple instances of this scope type simultaneously. When false, the runtime implementation MUST produce an error and stop deployment if an attempt is made to place a component into more than one instance of this scope type simultaneously. |
| `definitionRef` | [`DefinitionRef`](#definitionRef) | Y | | Index of the scope definition. |

### DefinitionRef

> We use the same [DefinitionRef](3.workload.md#DefinitionRef) in workload.

Reviewer 3: ```suggestion
We use the same [DefinitionRef](3.workload.md#DefinitionRef) in _WorkloadDefinition_.
```

> ## Core application scopes

Reviewer 4: I suggest we don't define core application scopes for now? Either Network Scope or Health Scope does not sound like a MUST implement to me. Maybe move them to standard scope @ryanzhang-oss

Author: totally agree

Reviewer: agree

> 

Reviewer: Will we still be having health/network scopes as core scopes? Or will scopes be selfimplemented CRD concepts(extensibles)?

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `allowComponentOverlap` | `bool` | Y | | Determines whether a component is allowed to be in multiple instances of this scope type simultaneously. When false, the runtime implementation MUST produce an error and stop deployment if an attempt is made to place a component into more than one instance of this scope type simultaneously. |
| `definitionRef` | [`DefinitionRef`](#definitionRef) | Y | | Index of the scope definition. |

> ### DefinitionRef

Reviewer 2: You can probably omit this and put `[DefinitionRef](3.workload.md#DefinitionRef)` at line 69

We use the same [DefinitionRef](3.workload.md#DefinitionRef) in workload.

## Deployment

> Application scope instances are defined and deployed by their standalone way. Go to scopes folder to see [core scopes](scopes/core.md) and [extended scopes](scopes/extended.md) details.

Reviewer 2: What does "Application scope instances are defined and deployed by their standalone way" mean?

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `allowComponentOverlap` | `bool` | Y | | Determines whether a component is allowed to be in multiple instances of this scope type simultaneously. When false, the runtime implementation MUST produce an error and stop deployment if an attempt is made to place a component into more than one instance of this scope type simultaneously. |
| `definitionRef` | [`DefinitionRef`](#definitionRef) | Y | | Index of the scope definition. |

### DefinitionRef

| Attribute | Type | Required | Default Value | Description |
|-----------|------|----------|---------------|-------------|
| `name` | `string` | Y | | The name of this scope definition. |
| `version` | `string` | Y | | The version of this scope definition. |

> | `schematic` | `string` | N | | The schematic of this scope properties. The value is a [JSON schema](https://json-schema.org/). |

Author: We just use one, I'll change it

A scope definitionRef __MUST__ must be unique by `name` and `version`. The `schematic` field is optional as the definitionRef could be simply an indexing for a schema. In this case a runtime itself __SHOULD__ store definitions and a definition __MUST__ contain a schematic which can validate scope spec.

## Deployment
Application scope instances are defined and deployed by their standalone way.

> ## Core scope type definitions

Reviewer 2: Can we also move all the definitions of core scopes to a separate place?

Author: removed

```yaml
apiVersion: core.oam.dev/v1alpha1
kind: ScopeDefinition
metadata:
  name: networkscope.core.oam.dev
spec:
  allowComponentOverlap: true
  definitionRef:

>      name: networkscope.core.oam.dev

Author: yeah， it's a ref same with trait defitionRef

```yaml
apiVersion: core.oam.dev/v1alpha1
kind: ScopeDefinition
metadata:
  name: healthscope.core.oam.dev
spec:
  allowComponentOverlap: true
  definitionRef:
    name: healthscope.core.oam.dev
    version: v1alpha1

>     schematic: |

Reviewer 3: According to https://github.com/oam-dev/spec/pull/304/, the scope definitionRef will point to a CRD name. We don't need to embed json schema here, do we?
